projects/storefrontlib/src/cms-components/myaccount/my-coupons/my-coupons.component.ts

```diff
@@ -0,0 +1,141 @@
+import { Component, OnInit } from '@angular/core';
+import { Observable, combineLatest } from 'rxjs';
+import {
+  CustomerCouponService,
+  CustomerCouponSearchResult,
+  PaginationModel,
+  TranslationService,
+} from '@spartacus/core';
+import { tap, map } from 'rxjs/operators';
+
+@Component({
+  selector: 'cx-my-coupons',
+  templateUrl: './my-coupons.component.html',
+})
+export class MyCouponsComponent implements OnInit {
+  couponResult$: Observable<CustomerCouponSearchResult>;
+  couponsLoading$: Observable<boolean>;
+  couponSubscriptionLoading$: Observable<boolean>;
+
+  private PAGE_SIZE = 10;
+  private sortMapping = {
+    byStartDateAsc: 'startDate:asc',
+    byStartDateDesc: 'startDate:desc',
+    byEndDateAsc: 'endDate:asc',
+    byEndDateDesc: 'endDate:desc',
+  };
+  sort = 'byStartDateAsc';
+
+  sortOptions = [
+    {
+      code: 'byStartDateAsc',
+      selected: false,
+    },
+    {
+      code: 'byStartDateDesc',
+      selected: false,
+    },
+    {
+      code: 'byEndDateAsc',
+      selected: false,
+    },
+    {
+      code: 'byEndDateDesc',
+      selected: false,
+    },
+  ];
+
+  pagination: PaginationModel;
+
+  constructor(
+    private couponService: CustomerCouponService,
+    private translation: TranslationService
+  ) {}
+
+  ngOnInit(): void {
+    this.couponService.loadCustomerCoupons(this.PAGE_SIZE);
+    this.couponResult$ = this.couponService
+      .getCustomerCoupons(this.PAGE_SIZE)
+      .pipe(
+        tap(
+          coupons =>
+            (this.pagination = {
+              currentPage: coupons.pagination.page,
+              pageSize: coupons.pagination.count,
+              totalPages: coupons.pagination.totalPages,
+              totalResults: coupons.pagination.totalCount,
+              sort: this.sort,
+            })
+        )
+      );
+    this.couponsLoading$ = this.couponService.getCustomerCouponsLoading();
+    this.couponSubscriptionLoading$ = combineLatest([
+      this.couponService.getSubscribeCustomerCouponResultLoading(),
+      this.couponService.getUnsubscribeCustomerCouponResultLoading(),
+    ]).pipe(
+      map(([subscribing, unsubscribing]) => subscribing || unsubscribing)
+    );
+  }
+
+  getSortLabels(): Observable<{
+    byStartDateAsc: string;
+    byStartDateDesc: string;
+    byEndDateAsc: string;
+    byEndDateDesc: string;
```

Comment:
This should be a property, if it's used in a template. Function will be recalculaten in each change detection cycle, and there will be new observable created each time.

Reply:
Resolved.